dfox-lib/src/db/sqlite.rs

```diff
@@ -0,0 +1,356 @@
+                            },
+                        };
+
+                        (column_name.to_string(), value)
+                    })
+                    .collect();
+
+                Value::Object(json_map)
+            })
+            .collect();
+
+        Ok(results)
+    }
+
+    async fn begin_transaction<'a>(&'a self) -> Result<Box<dyn Transaction + 'a>, DbError> {
+        let tx = self
+            .pool
+            .begin()
+            .await
+            .map_err(|e| DbError::Transaction(e.to_string()))?;
+        Ok(Box::new(SqliteTransaction { tx }))
+    }
+
+    async fn list_databases(&self) -> Result<Vec<String>, DbError> {
+        // SQLite doesn't support listing databases as it works with a single database file
+        Ok(vec!["main".to_string()])
+    }
+
+    async fn list_tables(&self) -> Result<Vec<String>, DbError> {
+        let query = r#"
+            SELECT name
+            FROM sqlite_master
+            WHERE type = 'table'
+        "#;
+
+        let rows = sqlx::query(query)
+            .fetch_all(&self.pool)
+            .await
+            .map_err(DbError::Sqlx)?;
+
+        let tables = rows
+            .iter()
+            .map(|row| row.try_get::<String, _>("name").unwrap_or_default())
+            .collect();
+
+        Ok(tables)
+    }
+
+    async fn describe_table(&self, table_name: &str) -> Result<TableSchema, DbError> {
+        let query = format!("PRAGMA table_info('{}')", table_name);
+        let rows = sqlx::query(&query)
+            .fetch_all(&self.pool)
+            .await
+            .map_err(DbError::Sqlx)?;
+
+        let columns = rows
+            .iter()
+            .map(|row| ColumnSchema {
+                name: row.try_get("name").unwrap(),
+                data_type: row.try_get("type").unwrap(),
+                is_nullable: row.try_get::<i64, _>("notnull").unwrap() == 0,
+                default: row.try_get("dflt_value").ok(),
+            })
+            .collect();
+
+        Ok(TableSchema {
+            table_name: table_name.to_string(),
+            columns,
+            indexes: Vec::new(),
+        })
+    }
+}
+
+pub struct SqliteTransaction<'a> {
+    tx: sqlx::Transaction<'a, Sqlite>,
+}
+
+#[async_trait]
+impl<'a> Transaction for SqliteTransaction<'a> {
+    async fn execute_transaction(&mut self, query: &str) -> Result<(), DbError> {
+        sqlx::query(query)
+            .execute(&mut *self.tx)
+            .await
+            .map_err(|e| DbError::Transaction(e.to_string()))?;
+        Ok(())
+    }
+
+    async fn commit_transaction(self: Box<Self>) -> Result<(), DbError> {
+        self.tx
+            .commit()
+            .await
+            .map_err(|e| DbError::Transaction(e.to_string()))
+    }
+
+    async fn rollback_transaction(self: Box<Self>) -> Result<(), DbError> {
+        self.tx
+            .rollback()
+            .await
+            .map_err(|e| DbError::Transaction(e.to_string()))
+    }
+}
```

Comment:
These implementations feel repetitive, aren't they?

That's most likely due to all of them using common `Database` trait from `sqlx`.

Design of yours still makes sense, as it allows to potentially implement connection to a SQL dialiect unknown to sqlx.

If you wish to avoid boilerplate, you can define a trait that has `sqlx::Database` as a supertrait, and only contains things that _differ_ among dialects. Then you can write a single generic implementation, and the code will be monomorphised for you.